presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/JdbcMetadata.java

```diff
@@ -302,6 +304,29 @@ public JdbcTableHandle getTableHandle(ConnectorSession session, SchemaTableName
         return Optional.of(new LimitApplicationResult<>(handle, jdbcClient.isLimitGuaranteed(session)));
     }
 
+    @Override
+    public Optional<TableScanRedirectApplicationResult> applyTableScanRedirect(ConnectorSession session, ConnectorTableHandle table)
+    {
+        JdbcTableHandle tableHandle = (JdbcTableHandle) table;
+
+        List<JdbcColumnHandle> sourceColumns = tableHandle.getColumns()
+                .map(columnSet -> (List<JdbcColumnHandle>) columnSet.stream().map(JdbcColumnHandle.class::cast).collect(toImmutableList()))
+                .orElseGet(() -> jdbcClient.getColumns(session, tableHandle));
+
+        Optional<JdbcTableScanRedirectionResult> result = jdbcClient.getRedirectedTableScan(session, tableHandle, sourceColumns);
+        if (result.isEmpty()) {
+            return Optional.empty();
+        }
+
+        Map<ColumnHandle, String> columnsMapping = result.get().getDestinationColumns();
+        return Optional.of(
+                new TableScanRedirectApplicationResult(
+                        result.get().getDestinationTable(),
+                        columnsMapping,
+                        tableHandle.getConstraint()
```

Comment:
@martint this is where `filter` in redirection is being used

Reply:
`filter` is removed from SPI now
The redirection rule now relies on `TableScanNode#getEnforcedConstraint` instead

Reply:
The rewrite shouldn't include any filters. The retargeted table should encompass all the filters and operations represented by the TableHandle that got redirected. So, the idea is, for instance:

Given a plan such as:

```
Aggregation -> Project -> Filter -> Scan(TH0)
```

the "pushdown into connector rules" might transform it into:

```
Aggregation -> Scan (TH1)
```

Where `TH1` represents a "virtual table" that's the result of scanning, filtering and projecting the rows.

Then, the retargeting API would detect that it knows about a table in some other connector that has data that matches the results of `TH1`, and allows the engine to rewrite the query into:

```
Aggregation -> Scan(TH2)
```

where `TH2` is a table in a different connector.

Reply:
@martint target table might be wider than pushed down predicate. For example,
target table might contain dates from 1990-2020, yet pushed down predicate might select dates >=2010. Therefore, we need to reapply enforced constraint and try to push it down into redirected table too.

Reply:
Fair enough. In that case, I have concerns about the current approach that relies on "enforced predicate". 

From a connector API perspective, the call says "given this handle to a virtual table, provide me with *something* that replaces it". The information included should contain everything that makes the replacement semantically equivalent. The "enforced predicate" is not part of this API -- it's just something that the planner "remembered" by storing it in the plan node associated with the table. Without that piece of information, the rewrite is incomplete and incorrect.

Reply:
@martint the initial API did include a `getFilter` in `TableScanRedirectApplicationResult`. We changed it later on to remove that and rely on enforced constraint from table scan node instead as it resulted in simpler code. Would you prefer that we just go back to returning filter through `TableScanRedirectApplicationResult` even if it's the same piece of information as the one in `TableScanNode#getEnforcedConstraint` ?

Reply:
As @raunaqmorarka mentioned this was an initial approach. However, the redirection rule should try to translate `enforcedConstraint` anyway, so returning the same constraint back from connector was redundant.

Maybe we should document in SPI doc that redirected table does not have to enforce pushed down constraints? One difference this and other pushdowns is that this new SPI do not return `ConnectorTableHandle`, so it's kind of unique.

Reply:
The SPI should stand on its own. It can't depend on the caller keeping track of things on its own for correctness. `enforcedConstraint` exists for convenience and it may not exist in the future.

Reply:
Ok. So I guess we should bring back TupleDomain<String> filter as part of redirection result then

Reply:
I've updated the PR with the change to bring back `TupleDomain<String> filter` as part of redirection result and rebased to master.